File: somsingphim/src/types/order.ts
// Domain TypeScript Contracts aligned strictly with Go Backend Domain Models (backend/internal/domain/)

export type OrderStatus =
  | 'QUOTATION'
  | 'PENDING_PAYMENT'
  | 'ORDER_CREATED'
  | 'FILE_CONFIRMED'
  | 'IN_PRODUCTION'
  | 'COMPLETED'
  | 'CANCELLED';

export type BindingType =
  | 'NONE'
  | 'PERFECT_HOT_GLUE'
  | 'SADDLE_STITCH'
  | 'WIRE_O'
  | 'PLASTIC_COMB'
  | 'CALENDAR';

export type ProductionStep =
  | 'PENDING'
  | 'PREPRESS_CHECK'
  | 'INNER_PRINTED'
  | 'COVER_PRINTED'
  | 'COVER_LAMINATED'
  | 'PAPER_TRIMMED'
  | 'BOUND'
  | 'READY_FOR_PICKUP'
  | 'COMPLETED';

export interface CustomPrintSpecs {
  size: string;
  paper: string;
  finishing?: string;
  lamination?: string;
  binding?: string;
  width_mm?: number | string;
  height_mm?: number | string;
  width_cm?: number;
  height_cm?: number;
  pages?: number;
  is_double_sided?: boolean;
  grommets_count?: number;
  edge_folding?: boolean;
  ink_coverage_percent?: number | string;
  ink_coverage_c?: number | string;
  ink_coverage_m?: number | string;
  ink_coverage_y?: number | string;
  ink_coverage_k?: number | string;
  color_mode?: string;
  spoilage_rate_pct?: number | string;
  additional_notes?: string;
}

export type PrintSpecification = CustomPrintSpecs;

export interface CoverageMetrics {
  c: number | string;
  m: number | string;
  y: number | string;
  k: number | string;
  tac: number | string;
}

export interface InternalOrderPricing {
  base_material_cost_lak: number;
  ink_usage_cost_lak: number;
  plate_cost_lak: number;
  machine_depreciation_lak: number;
  labor_finishing_cost_lak: number;
  waste_spoilage_cost_lak: number;
  net_internal_cost_lak: number;
  markup_amount_lak: number;
  tax_amount_lak: number;
  total_price_lak: number;
  unit_price_lak: number;
  genuine_ink_baseline_lak?: number;
  compatible_ink_cost_lak?: number;
  ink_savings_lak?: number;
  ink_savings_percent?: number;
}

export type CostBreakdown = InternalOrderPricing;
export type AdminPricingBreakdown = InternalOrderPricing;

export interface OrderItemCostBreakdown {
  id: string;
  order_item_id: string;
  paper_cost: string | number;
  ink_cost: string | number;
  binding_cost: string | number;
  finishing_cost: string | number;
  unit_price: string | number;
  total_price: string | number;
  raw_c_pct: string | number;
  raw_m_pct: string | number;
  raw_y_pct: string | number;
  raw_k_pct: string | number;
  raw_tac_pct: string | number;
  applied_tac_pct: string | number;
  is_manual_override: boolean;
  created_at: string;
}

export interface OrderStatusHistory {
  id: string;
  order_id: string;
  previous_status: OrderStatus;
  new_status: OrderStatus;
  reason?: string;
  performed_by?: string;
  created_at: string;
}

export interface SpoilageLog {
  id: string;
  order_id: string;
  order_item_id?: string;
  material_sku: string;
  material_name: string;
  category: string;
  quantity_spoiled: string | number;
  unit: string;
  reason: string;
  cost_impact: string | number;
  recorded_by?: string;
  created_at: string;
}

export interface OrderItem {
  id: string;
  order_id: string;
  product_id?: string;
  job_name: string;
  item_name: string;
  quantity: number;
  page_count: number;
  paper_size: string;
  paper_sku?: string;
  binding_type: BindingType;
  spine_width_mm: string | number;
  unit_price: string | number;
  unit_cost: string | number;
  total_price: string | number;
  total_cost: string | number;
  is_manual_override: boolean;
  override_reason?: string;
  override_by?: string;
  specs: CustomPrintSpecs;
  cost_breakdown?: InternalOrderPricing;
  created_at: string;
  updated_at: string;
}

export interface Order {
  id: string;
  order_number: string;
  customer_id?: string;
  customer_name: string;
  customer_phone?: string;
  customer_email?: string;
  customer_address?: string;
  status: OrderStatus;
  total_amount: string | number;
  deposit_amount: string | number;
  remaining_amount: string | number;
  currency: string;
  exchange_rate: string | number;
  google_drive_link?: string;
  proof_url?: string;
  proof_approved_at?: string | null;
  proof_rejected_at?: string | null;
  proof_rejection_reason?: string;
  stock_deducted_at?: string | null;
  delivery_date?: string;
  notes?: string;
  created_by?: string;
  items: OrderItem[];
  status_histories?: OrderStatusHistory[];
  spoilage_logs?: SpoilageLog[];
  created_at: string;
  updated_at: string;
}

export interface ProductPricingTemplate {
  id: string;
  name: string;
  material_id: string;
  baseline_coverage_percent: string | number;
  coverage_surcharge_multiplier: string | number;
  min_order_quantity: number;
  min_total_price: string | number;
  addon_rates: Record<string, any>;
  is_active: boolean;
  created_at: string;
  updated_at: string;
}

export interface CreatePricingTemplatePayload {
  name: string;
  material_id: string;
  baseline_coverage_percent?: string | number;
  coverage_surcharge_multiplier?: string | number;
  min_order_quantity?: number;
  min_total_price?: string | number;
  addon_rates?: Record<string, any>;
  is_active?: boolean;
}

export interface UpdatePricingTemplatePayload {
  name?: string;
  material_id?: string;
  baseline_coverage_percent?: string | number;
  coverage_surcharge_multiplier?: string | number;
  min_order_quantity?: number;
  min_total_price?: string | number;
  addon_rates?: Record<string, any>;
  is_active?: boolean;
}

export interface AddonItemBreakdown {
  name: string;
  type: string;
  quantity: string | number;
  unit_price: string | number;
  total_cost: string | number;
}

export interface DynamicPriceBreakdown {
  template_id: string;
  template_name: string;
  material_id: string;
  material_name: string;
  quantity: number;
  min_order_quantity: number;
  area_m2: string | number;
  perimeter_m: string | number;
  base_unit_price: string | number;
  base_material_cost: string | number;
  baseline_coverage_percent: string | number;
  actual_coverage_percent: string | number;
  coverage_delta_percent: string | number;
  coverage_surcharge_multiplier: string | number;
  coverage_surcharge: string | number;
  addon_cost: string | number;
  itemized_addons: AddonItemBreakdown[];
  subtotal: string | number;
  min_total_price: string | number;
  min_price_applied: boolean;
  final_price: string | number;
  final_unit_price: string | number;
}

export interface PricingCalculationRequest {
  job_name: string;
  quantity: number;
  template_id?: string;
  paper_sku?: string;
  paper_cost_per_unit_lak?: number;
  paper_format?: string;
  sheets_per_pack?: number;
  cuts_per_sheet?: number;
  width_cm?: number;
  height_cm?: number;
  unfolded_width_mm?: number;
  unfolded_height_mm?: number;
  page_count?: number;
  ink_coverage_percent?: number;
  ink_cost_per_ml_lak?: number;
  use_compatible_ink?: boolean;
  compatible_ink_cost_per_ml_lak?: number;
  lamination_type?: string;
  lamination_cost_lak?: number;
  binding_type?: string;
  binding_cost_lak?: number;
  grommets_count?: number;
  grommet_cost_lak?: number;
  edge_folding?: boolean;
  folding_cost_lak?: number;
  labor_hours?: number;
  labor_rate_per_hour_lak?: number;
  machine_depreciation_rate_lak?: number;
  plate_cost_per_unit_lak?: number;
  spoilage_rate_percent?: number;
  markup_margin_percent?: number;
  tax_rate_percent?: number;
  min_total_price_lak?: number;
}

export interface PricingCalculationResponse {
  job_name: string;
  quantity: number;
  unit_price_lak: number;
  total_price_lak: number;
  cost_breakdown: InternalOrderPricing;
  currency: string;
}

export type PricingResponse = PricingCalculationResponse;

File: backend/internal/domain/order.go
package domain

import (
	"errors"
	"time"

	"github.com/shopspring/decimal"
)

// OrderStatus defines the strict state machine lifecycle of print orders
type OrderStatus string

const (
	StatusQuotation      OrderStatus = "QUOTATION"
	StatusPendingPayment OrderStatus = "PENDING_PAYMENT"
	StatusOrderCreated   OrderStatus = "ORDER_CREATED"
	StatusFileConfirmed  OrderStatus = "FILE_CONFIRMED"
	StatusInProduction   OrderStatus = "IN_PRODUCTION"
	StatusCompleted      OrderStatus = "COMPLETED"
	StatusCancelled      OrderStatus = "CANCELLED"
)

// BindingType defines available bookbinding/finishing methods
type BindingType string

const (
	BindingNone           BindingType = "NONE"
	BindingPerfectHotGlue BindingType = "PERFECT_HOT_GLUE"
	BindingSaddleStitch   BindingType = "SADDLE_STITCH"
	BindingWireO          BindingType = "WIRE_O"
	BindingPlasticComb    BindingType = "PLASTIC_COMB"
	BindingCalendar       BindingType = "CALENDAR"
)

var (
	ErrInvalidStatusTransition = errors.New("invalid order status transition")
	ErrOrderAlreadyCompleted    = errors.New("cannot change status of completed order")
	ErrOrderAlreadyCancelled    = errors.New("cannot change status of cancelled order")
	ErrStockAlreadyDeducted     = errors.New("inventory stock already deducted for this order")
)

// AllowedTransitions maps valid forward and backward status transitions
var AllowedTransitions = map[OrderStatus][]OrderStatus{
	StatusQuotation:      {StatusPendingPayment, StatusCancelled},
	StatusPendingPayment: {StatusOrderCreated, StatusQuotation, StatusCancelled},
	StatusOrderCreated:   {StatusFileConfirmed, StatusCancelled},
	StatusFileConfirmed:  {StatusInProduction, StatusCancelled},
	StatusInProduction:   {StatusCompleted, StatusCancelled},
	StatusCompleted:      {},
	StatusCancelled:      {},
}

// ValidateStatusTransition checks if transition from current to target status is permitted
func ValidateStatusTransition(from, to OrderStatus) error {
	if from == to {
		return nil
	}
	if from == StatusCompleted {
		return ErrOrderAlreadyCompleted
	}
	if from == StatusCancelled {
		return ErrOrderAlreadyCancelled
	}

	allowed, ok := AllowedTransitions[from]
	if !ok {
		return ErrInvalidStatusTransition
	}

	for _, s := range allowed {
		if s == to {
			return nil
		}
	}
	return ErrInvalidStatusTransition
}

// CustomPrintSpecs specifies detailed item technical configuration
type CustomPrintSpecs struct {
	Size               string          `json:"size"`
	Paper              string          `json:"paper"`
	Finishing          string          `json:"finishing,omitempty"`
	Lamination         string          `json:"lamination,omitempty"`
	Binding            string          `json:"binding,omitempty"`
	WidthMM            decimal.Decimal `json:"width_mm,omitempty"`
	HeightMM           decimal.Decimal `json:"height_mm,omitempty"`
	Pages              int             `json:"pages,omitempty"`
	IsDoubleSided      bool            `json:"is_double_sided,omitempty"`
	InkCoveragePercent decimal.Decimal `json:"ink_coverage_percent,omitempty"`
	InkCoverageC       decimal.Decimal `json:"ink_coverage_c,omitempty"`
	InkCoverageM       decimal.Decimal `json:"ink_coverage_m,omitempty"`
	InkCoverageY       decimal.Decimal `json:"ink_coverage_y,omitempty"`
	InkCoverageK       decimal.Decimal `json:"ink_coverage_k,omitempty"`
	ColorMode          string          `json:"color_mode,omitempty"`
	SpoilageRatePct    decimal.Decimal `json:"spoilage_rate_pct,omitempty"`
	AdditionalNotes    string          `json:"additional_notes,omitempty"`
}

// OrderItem represents a single product or book line item in an order
type OrderItem struct {
	ID               string                 `json:"id" db:"id"`
	OrderID          string                 `json:"order_id" db:"order_id"`
	ProductID        string                 `json:"product_id,omitempty" db:"product_id"`
	JobName          string                 `json:"job_name" db:"job_name"`
	ItemName         string                 `json:"item_name" db:"item_name"`
	Quantity         int                    `json:"quantity" db:"quantity"`
	PageCount        int                    `json:"page_count" db:"page_count"`
	PaperSize        string                 `json:"paper_size" db:"paper_size"`
	PaperSKU         string                 `json:"paper_sku,omitempty" db:"paper_sku"`
	BindingType      BindingType            `json:"binding_type" db:"binding_type"`
	SpineWidthMM     decimal.Decimal        `json:"spine_width_mm" db:"spine_width_mm"`
	UnitPrice        decimal.Decimal        `json:"unit_price" db:"unit_price"`
	UnitCost         decimal.Decimal        `json:"unit_cost" db:"unit_cost"`
	TotalPrice       decimal.Decimal        `json:"total_price" db:"total_price"`
	TotalCost        decimal.Decimal        `json:"total_cost" db:"total_cost"`
	IsManualOverride bool                   `json:"is_manual_override" db:"is_manual_override"`
	OverrideReason   string                 `json:"override_reason,omitempty" db:"override_reason"`
	OverrideBy       string                 `json:"override_by,omitempty" db:"override_by"`
	Specs            CustomPrintSpecs       `json:"specs" db:"specs"`
	CostBreakdown    *InternalCostAudit     `json:"cost_breakdown,omitempty"`
	CreatedAt        time.Time              `json:"created_at" db:"created_at"`
	UpdatedAt        time.Time              `json:"updated_at" db:"updated_at"`
}

// Order represents an authoritative print order entity
type Order struct {
	ID                    string                 `json:"id" db:"id"`
	OrderNumber           string                 `json:"order_number" db:"order_number"`
	CustomerID            string                 `json:"customer_id,omitempty" db:"customer_id"`
	CustomerName          string                 `json:"customer_name" db:"customer_name"`
	CustomerPhone         string                 `json:"customer_phone,omitempty" db:"customer_phone"`
	CustomerEmail         string                 `json:"customer_email,omitempty" db:"customer_email"`
	CustomerAddress       string                 `json:"customer_address,omitempty" db:"customer_address"`
	Status                OrderStatus            `json:"status" db:"status"`
	TotalAmount           decimal.Decimal        `json:"total_amount" db:"total_amount"`
	DepositAmount         decimal.Decimal        `json:"deposit_amount" db:"deposit_amount"`
	RemainingAmount       decimal.Decimal        `json:"remaining_amount" db:"remaining_amount"`
	Currency              string                 `json:"currency" db:"currency"`
	ExchangeRate          decimal.Decimal        `json:"exchange_rate" db:"exchange_rate"`
	GoogleDriveLink       string                 `json:"google_drive_link,omitempty" db:"google_drive_link"`
	ProofURL              string                 `json:"proof_url,omitempty" db:"proof_url"`
	ProofApprovedAt       *time.Time             `json:"proof_approved_at,omitempty" db:"proof_approved_at"`
	ProofRejectedAt       *time.Time             `json:"proof_rejected_at,omitempty" db:"proof_rejected_at"`
	ProofRejectionReason  string                 `json:"proof_rejection_reason,omitempty" db:"proof_rejection_reason"`
	StockDeductedAt       *time.Time             `json:"stock_deducted_at,omitempty" db:"stock_deducted_at"`
	IdempotencyKey        string                 `json:"idempotency_key,omitempty" db:"idempotency_key"`
	DeliveryDate          string                 `json:"delivery_date,omitempty" db:"delivery_date"`
	Notes                 string                 `json:"notes,omitempty" db:"notes"`
	CreatedBy             string                 `json:"created_by,omitempty" db:"created_by"`
	Items                 []OrderItem            `json:"items,omitempty"`
	StatusHistories       []OrderStatusHistory   `json:"status_histories,omitempty"`
	SpoilageLogs          []SpoilageLog          `json:"spoilage_logs,omitempty"`
	CreatedAt             time.Time              `json:"created_at" db:"created_at"`
	UpdatedAt             time.Time              `json:"updated_at" db:"updated_at"`
}

// OrderStatusHistory tracks every state transition and audit trail
type OrderStatusHistory struct {
	ID             string      `json:"id" db:"id"`
	OrderID        string      `json:"order_id" db:"order_id"`
	PreviousStatus OrderStatus `json:"previous_status" db:"previous_status"`
	NewStatus      OrderStatus `json:"new_status" db:"new_status"`
	Reason         string      `json:"reason,omitempty" db:"reason"`
	PerformedBy    string      `json:"performed_by,omitempty" db:"performed_by"`
	CreatedAt      time.Time   `json:"created_at" db:"created_at"`
}

// SpoilageLog tracks waste and spoilage during production
type SpoilageLog struct {
	ID              string          `json:"id" db:"id"`
	OrderID         string          `json:"order_id" db:"order_id"`
	OrderItemID     string          `json:"order_item_id,omitempty" db:"order_item_id"`
	MaterialSKU     string          `json:"material_sku" db:"material_sku"`
	MaterialName    string          `json:"material_name" db:"material_name"`
	Category        string          `json:"category" db:"category"`
	QuantitySpoiled decimal.Decimal `json:"quantity_spoiled" db:"quantity_spoiled"`
	Unit            string          `json:"unit" db:"unit"`
	Reason          string          `json:"reason" db:"reason"`
	CostImpact      decimal.Decimal `json:"cost_impact" db:"cost_impact"`
	RecordedBy      string          `json:"recorded_by,omitempty" db:"recorded_by"`
	CreatedAt       time.Time       `json:"created_at" db:"created_at"`
}

// CreateOrderItemPayload represents request payload for an item
type CreateOrderItemPayload struct {
	ProductID    string           `json:"product_id,omitempty"`
	JobName      string           `json:"job_name" binding:"required"`
	ItemName     string           `json:"item_name" binding:"required"`
	Quantity     int              `json:"quantity" binding:"required,gt=0"`
	PageCount    int              `json:"page_count" binding:"required,gt=0"`
	PaperSize    string           `json:"paper_size"`
	PaperSKU     string           `json:"paper_sku"`
	BindingType  BindingType      `json:"binding_type"`
	SpineWidthMM decimal.Decimal  `json:"spine_width_mm"`
	UnitPrice    decimal.Decimal  `json:"unit_price"`
	UnitCost     decimal.Decimal  `json:"unit_cost"`
	Specs        CustomPrintSpecs `json:"specs"`
}

// CreateOrderPayload represents request payload to create an order or quotation
type CreateOrderPayload struct {
	OrderNumber     string                   `json:"order_number"`
	CustomerID      string                   `json:"customer_id,omitempty"`
	CustomerName    string                   `json:"customer_name" binding:"required"`
	CustomerPhone   string                   `json:"customer_phone"`
	CustomerEmail   string                   `json:"customer_email"`
	CustomerAddress string                   `json:"customer_address"`
	Status          OrderStatus              `json:"status"` // Defaults to QUOTATION
	DepositAmount   decimal.Decimal          `json:"deposit_amount"`
	Currency        string                   `json:"currency"`
	ExchangeRate    decimal.Decimal          `json:"exchange_rate"`
	GoogleDriveLink string                   `json:"google_drive_link"`
	DeliveryDate    string                   `json:"delivery_date"`
	Notes           string                   `json:"notes"`
	CreatedBy       string                   `json:"created_by"`
	Items           []CreateOrderItemPayload `json:"items" binding:"required,min=1"`
}

// UpdateOrderStatusPayload represents request payload to transition status
type UpdateOrderStatusPayload struct {
	Status      OrderStatus `json:"status" binding:"required"`
	Reason      string      `json:"reason"`
	PerformedBy string      `json:"performed_by"`
}

// OverridePricingPayload represents request payload to override item pricing
type OverridePricingPayload struct {
	OrderItemID       string          `json:"order_item_id" binding:"required"`
	OverrideUnitPrice decimal.Decimal `json:"override_unit_price" binding:"required"`
	Reason            string          `json:"reason" binding:"required"`
	ApprovedBy        string          `json:"approved_by" binding:"required"`
}

// OrderFilter represents criteria for listing orders
type OrderFilter struct {
	Status       OrderStatus
	CustomerName string
	OrderNumber  string
	Limit        int
	Offset       int
}
